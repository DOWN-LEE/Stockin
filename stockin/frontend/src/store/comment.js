import { createSlice } from '@reduxjs/toolkit';
import { api } from '../api/index';

const initialState = {
  commentList: [],
};

const slice = createSlice({
  name: 'comment',
  initialState,
  reducers: {
    saveCommentList: (state, action) => {
      state.commentList = action.payload;
    },
  },
});

export default slice.reducer;

// Actions

const { saveCommentList } = slice.actions;
export const updateCommentList = (stock_id) => async (dispatch) => {
  try {
    await api
      .get('/stocks/' + stock_id + '/comments/')
      .then((response) => dispatch(saveCommentList(response.data)));
  } catch (e) {
    return console.error(e.message);
  }
};
export const commentList = state.comment.commentList;
