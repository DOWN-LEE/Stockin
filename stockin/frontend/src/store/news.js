import { createSlice } from '@reduxjs/toolkit';
import { toInteger } from 'lodash';
import { api } from '../api/index';

const slice = createSlice({
  name: 'news',
  initialState: {
    news : [],
  },
  reducers: {
    getnews: (state, action) => {
      console.log(action.payload);
      return {...state, news : action.payload }
    },
  },
});

export default slice.reducer;

// Actions

const { getnews } = slice.actions;
export const getNews = (id, date) => async (dispatch) => {
  try {
    await api.get('/news/stocks/'+id+'/date/'+date+'/').then((response) => dispatch(getnews(response.data)));
  } 
  catch (e) {
    return console.error(e.message);
  }
};
