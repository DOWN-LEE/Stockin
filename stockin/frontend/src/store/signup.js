import { createSlice } from '@reduxjs/toolkit';
import { api } from '../api/index';

const slice = createSlice({
  name: 'user',
  initialState: {
    userList: [],
  },
  reducers: {
    signup: (state, action) => {
      state.userList.push({
        email: action.payload.email,
        nickname: action.payload.nickname,
        password: action.payload.password,
      });
    },
  },
});

export default slice.reducer;

// Actions

const { signup } = slice.actions;

export const trySignup = (user) => async (dispatch) => {
  try {
    await api.post('/users/signup/', user).then((response) => dispatch(signup(response.data)));
  } catch (e) {
    return console.error(e.message);
  }
};
