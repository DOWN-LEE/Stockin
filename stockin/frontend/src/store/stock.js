import { createSlice } from '@reduxjs/toolkit';
import { api } from '../api/index';
import axios from 'axios';

const slice = createSlice({
  name: 'stock',
  initialState: {
    stockList : [],
  },
  reducers: {
    stocklist: (state, action) => {
      console.log("a");
      return {
        ...state,
        stockList : action.payload,
      }
    },
  },
});

export default slice.reducer;

// Actions
const { stocklist } = slice.actions;
export const getStocks = () => async (dispatch) => {
  try {
    await axios.get('http://localhost:8000/api/stocks/').then((response) => {
      dispatch(stocklist(response.data));
    })
  } catch (e) {
    return console.error(e.message);
  }
};


