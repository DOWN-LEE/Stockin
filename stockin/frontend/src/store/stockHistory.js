import { createSlice } from '@reduxjs/toolkit';
import { api } from '../api/index';
import { timeParse } from 'd3-time-format';

const slice = createSlice({
  name: 'stockHistory',
  initialState: {
    priceList: [],
  },
  reducers: {
    stockHistory: (state, action) => {
      let data = action.payload;
      let parseDate = timeParse('%Y-%m-%dT%H:%M:%SZ');
      data.forEach((d, i) => {
        d.date = parseDate(d.date);
        d.open = +d.open;
        d.high = +d.high;
        d.low = +d.low;
        d.close = +d.close;
      });
      state.priceList = data;
    },
  },
});

export default slice.reducer;

// Actions

const { stockHistory } = slice.actions;

export const getStockHistory = (stock_id) => async (dispatch) => {
  try {
    await api
      .get('/stocks/price/' + stock_id + '/', user)
      .then((response) => dispatch(stockHistory(response.data)));
  } catch (e) {
    return console.error(e.message);
  }
};
