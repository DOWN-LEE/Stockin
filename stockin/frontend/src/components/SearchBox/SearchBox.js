import _ from 'lodash';
import React from 'react';
import { Search, Grid, Header, Segment, Label } from 'semantic-ui-react';
import { useHistory } from "react-router";
import { useSelector, useDispatch } from 'react-redux';
import { getStocks } from '../../store/stock';

const keyGenerator = () => '_' + Math.random().toString(36).substr(2, 9);

const handleOnClick = (title) => {
  console.log('clicked : ' + {title});
  var retrieve_list = JSON.parse(localStorage.getItem('recent-search')) || [];
  console.log(retrieve_list);
  const idx = retrieve_list.findIndex(function(elem) {return elem.title === title});
  if (idx > -1) {
    retrieve_list.splice(idx, 1);
  }
  localStorage.setItem('recent-search', JSON.stringify(retrieve_list));
  console.log(retrieve_list);
};
const resultRenderer = ({ title, description }) => (
  <div>
    <p>{title}</p>
    <p>{description}</p>
    <button onclick={handleOnClick(title)}>button</button>
  </div>
);

const initialState = {
  loading: false,
  results: [],
  value: '',
  stocks: [],
};

function exampleReducer(state, action) {
  switch (action.type) {
    case 'CLEAN_QUERY':
      return initialState;
    case 'START_SEARCH':
      return { ...state, loading: true, value: action.query };
    case 'FINISH_SEARCH':
      const fin_results = (JSON.parse(localStorage.getItem('recent-search')) || []).concat(action.results);
      return { ...state, loading: false, results: fin_results };
    case 'UPDATE_SELECTION':
      return { ...state, value: action.selection };

    default:
      throw new Error();
  }
}

function SearchBox() {
  const history = useHistory(); 
  const _dispatch = useDispatch();
  const { stockList } = useSelector((state) => state.stock);
  const [state, dispatch] = React.useReducer(exampleReducer, initialState);
  const { loading, results, value } = state;
  const timeoutRef = React.useRef();

  React.useEffect(() => {
    state.results = JSON.parse(localStorage.getItem('recent-search')) || [];
    _dispatch(getStocks());
    return () => {
      clearTimeout(timeoutRef.current);
    };
  }, []);

  const handleSearchChange = React.useCallback((e, data) => {
    clearTimeout(timeoutRef.current);
    dispatch({ type: 'START_SEARCH', query: data.value });

    timeoutRef.current = setTimeout(() => {
      if (data.value.length === 0) {
        dispatch({ type: 'CLEAN_QUERY' });
        return;
      }
      var search_result = [];
      var _search_result = stockList.filter((st) => st.title.toUpperCase().includes(data.value.toUpperCase())).slice(0,5);
      
      var i;
      for(i = 0; i < _search_result.length; i++) {
        var temp = {..._search_result[i], key : keyGenerator(), description : _search_result[i].sector, };
        search_result.push(temp);
      }

      dispatch({type: 'FINISH_SEARCH', results: search_result});
    }, 300);
  }, [stockList]);

  const handleResultSelect = (e, data) => {
    var selected_stock = data.result;
    var retrieve_list = JSON.parse(localStorage.getItem('recent-search')) || [];
    
    if(retrieve_list.length === 0
       || !retrieve_list.find(element => element.id === selected_stock.id)) {
      selected_stock.key = keyGenerator();
      selected_stock = {...selected_stock, key : keyGenerator(), description : "최근 검색" }
      retrieve_list.push(selected_stock);
      localStorage.setItem('recent-search', JSON.stringify(retrieve_list));
    }
  
    dispatch({ type: 'UPDATE_SELECTION', selection: selected_stock.title });
    history.push('/detail/' + data.result.id)
  }

  return (
    <Search
      data-testid="SearchBox"
      fluid
      action
      loading={loading}
      onSearchChange={handleSearchChange}
      onResultSelect={handleResultSelect}
      resultRenderer={resultRenderer}
      results={results}
      value={value}
    />
  );
}

export default SearchBox;
