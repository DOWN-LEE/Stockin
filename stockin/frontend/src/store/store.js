import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import authentication from './authentication/authentication';
import signup from './signup';
import comment from './comment';
import stock from './stock';
import news from './news';
import financialstats from './financialstats';
import stockHistory from './stockHistory';

export const history = createBrowserHistory();

const rootReducer = combineReducers({
  router: connectRouter(history),
  fs: financialstats,
  authentication,
  signup,
  comment,
  stock,
  news,
  stockHistory,
});
export const middlewares = [thunk, routerMiddleware(history)];

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(rootReducer, composeEnhancers(applyMiddleware(...middlewares)));

export default store;
