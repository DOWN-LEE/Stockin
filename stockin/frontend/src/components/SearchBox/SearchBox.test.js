import React from 'react';
import {
  render,
  screen,
  fireEvent,
  getByText,
  queryAllByTestId,
  getByPlaceholderText,
} from '@testing-library/react';
import SearchBox from './SearchBox';
import store from '../../store/store';
import { Provider } from 'react-redux';
import { Container, Menu, Button, Image, Visibility } from 'semantic-ui-react';
import { head } from 'lodash';

describe('<SearchBox />', () => {
  let searchbox;
  beforeEach(() => {
    searchbox = (
      <Provider store={store}>
        <SearchBox/>
      </Provider>
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should handle value change', () => {
    const utils = render(searchbox);
    const input = document.body.querySelector('.prompt');
    fireEvent.change(input, { target: { value: 'foo' } })
    expect(input.value).toBe('foo') 
    fireEvent.change(input, { target: { value: '' } })
    console.log(input.value);
    expect(input.value).toBe('')
  });

  it('should represent correct ', () => {
    const utils = render(searchbox);
    const input = document.body.querySelector('.prompt');
    fireEvent.change(input, { target: { value: 'foo' } });
    expect(input.value).toBe('foo');
    const query = screen.queryAllByText(/No results found./i);
    expect(query.length).toBe(1);
  });


});