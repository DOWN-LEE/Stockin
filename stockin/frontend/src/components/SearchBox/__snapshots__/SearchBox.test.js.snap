// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> should render without errors 1`] = `
<div>
  <div
    class="ui fluid search"
    data-testid="SearchBox"
  >
    <div
      class="ui icon input"
    >
      <input
        autocomplete="off"
        class="prompt"
        tabindex="0"
        type="text"
        value=""
      />
      <i
        aria-hidden="true"
        class="search icon"
      />
    </div>
    <div
      class="results transition"
    >
      <div
        class="message empty"
      >
        <div
          class="header"
        >
          No results found.
        </div>
      </div>
    </div>
  </div>
</div>
`;
